libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +930,70 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type, bool _fromCalldata, bool _padToWords)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&] {
+		if (auto arrayType = dynamic_cast<ArrayType const*>(&_type))
+		{
+			solAssert(!arrayType->isDynamicallySized(), "");
+			solAssert(!_fromCalldata, "");
+			solAssert(_padToWords, "");
+
+			return Whiskers(R"(
+				function <functionName>(addr) -> value {
+					value := addr
+				}
+			)")
+			("functionName", functionName)
+			.render();
+		}
+
+		unsigned numBytes = _type.calldataEncodedSize(_padToWords);
+
+		solAssert(numBytes != 0, "");
+
+		if (auto const* funType = dynamic_cast<FunctionType const*>(&_type))
+			if (funType->kind() == FunctionType::Kind::External)
+				return Whiskers(R"(
+					function <functionName>(addr) -> addr, selector {
+						addr := <load>(addr)
+						addr, selector := <splitFunction>(addr)
+						<?needsValidation>
+							selector := <validate>(selector)
```

Comment:
Only the selector?

Reply:
I belive the original code does it like that: https://github.com/ethereum/solidity/blob/develop/libsolidity/codegen/CompilerUtils.cpp#L1366-L1383 but I can validate both of course. Will change.

Reply:
Actually, isn't this a problem with the type? Because the type the validate function has is for `_type`, but `addr` and `selector` are different from type, correct?

Reply:
Same problem would apply in the old code

Reply:
The old code does not properly validate everything. `addr` and `selector` are the two values associated with the `function` type. The function type occupies two stack slots which translates to two variables in yul. It gets a bit more complicated because sometimes, they are combined into a single value. I just checked, and it turns out that the validation function is not written yet for non-combined external function types. So maybe better to leave this out for now.